[CBR-374] V2 error handling preparations in WalletLayer and Kernel

Anviking opened the pull request:
## Description
This PR introduces a root exception type `WalletException` (as described in [Control.Exception](http://hackage.haskell.org/package/base-4.11.1.0/docs/Control-Exception.html)) that, in the future *could* be used like this:
```haskell
instance Exception CreateAccountError where
    toException   = walletExceptionToException
    fromException = walletExceptionFromException
```
and allow catching all WalletExceptions with one handler.

It also adds `Arbitrary` and `Aeson` instances for the inherent types of the WalletLayer errors.

### New plan
- [x] Drop obvious commits
- [x] Ensure it compiles and passes tests
- [x] Split *[CBR-374] Fix bad arbitrary instances* (separate Aeson changes)
- [x] Squash *Fixup simplify Aeson import*
- [x] Squash all other fixups
- [ ] <s>Split/reword *Fix bad Arbitrary instances* (also contains new instances)</s>
- [x] Rebase from current develop
- [x] Comment `WalletLayer.Ecxeption`

## Linked issue

[CBR-374](https://iohk.myjetbrains.com/youtrack/issue/CBR-374)

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

Anviking commented on wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs:
```diff
@@ -14,10 +14,15 @@ import qualified Formatting as F
 import qualified Formatting.Buildable
 import           System.Random.MWC (GenIO, createSystemRandom, uniformR)
 
+import           Servant (err400)
```
Maybe it's ugly to import all of the things required for the instances in each file containing errors.

Anviking commented on wallet-new/src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -161,7 +162,7 @@ legacyWalletBackend pm txpConfig WalletBackendParams {..} ntpStatus = pure $ \di
     exceptionHandler :: SomeException -> Response
     exceptionHandler se =
-        case asum [handleV1Errors se, handleV0Errors se] of
+        case asum [handleV1Errors se, handleWalletException <$> fromException se, handleV0Errors se] of
```
Different styles here, I like it better though

Could make all handlers `e -> Response` too.

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs:
```diff
@@ -0,0 +1,42 @@
+instance Buildable WalletException where
+    build (WalletException e) = build e
+
+instance Exception WalletException
```
I don't think we need all those instances. Only `Show` is required to define the `Exception`.  So it could basically be:

```hs
data WalletException = forall e.
    ( ToHttpErrorStatus e
    , HasDiagnostic e
    , Exception e
    ) => WalletException { getWalletException :: e }

instance Prelude.Show WalletException where
    show (WalletException e) = Prelude.show e

instance Exception WalletException
```

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs:
```diff
@@ -58,10 +65,18 @@ instance Buildable CreateAccountError where
-instance Exception CreateAccountError
+instance Exception CreateAccountError where
+    toException   = walletExceptionToException
+    fromException = walletExceptionFromException
```
Note that this need to go in  `Cardano/Wallet//WalletLayer.hs` we aren't expecting the Kernel to throw anything :smile:

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs:
```diff
@@ -0,0 +1,42 @@
+walletExceptionFromException :: Exception e => SomeException -> Maybe e
+walletExceptionFromException x = do
+    WalletException a <- fromException x
+    cast a
```
I'd suggest to trade this for:

```hs
throwWalletM ::
    ( ToHttpErrorStatus e
    , HasDiagnostic e
    , Exception e
    , MonadThrow m
    ) => e -> m a
throwWalletM =
    throwM . WalletException
```

And use this `throwWalletM` in the different handlers :+1:

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs:
```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Cardano.Wallet.Kernel.WalletException where
```
This need to be in `Cardano.Wallet.WalletLayer.Exception` :+1:

KtorZ commented:
Since we are a bit pressured by time, here's a patch with my comments:

```
diff --git a/wallet-new/src/Cardano/Wallet/API/V1/Generic.hs b/wallet-new/src/Cardano/Wallet/API/V1/Generic.hs
index d6e0acb..2dfa883 100644
--- a/wallet-new/src/Cardano/Wallet/API/V1/Generic.hs
+++ b/wallet-new/src/Cardano/Wallet/API/V1/Generic.hs
@@ -5,6 +5,7 @@ module Cardano.Wallet.API.V1.Generic
     , jsendErrorGenericParseJSON
     , gconsNames
     , gconsName
+    , GDiagnosticToJSON
     ) where
 
 import           Universum hiding (All, Generic)
diff --git a/wallet-new/src/Cardano/Wallet/API/V1/Handlers/Accounts.hs b/wallet-new/src/Cardano/Wallet/API/V1/Handlers/Accounts.hs
index a7523fc..eb18bf0 100644
--- a/wallet-new/src/Cardano/Wallet/API/V1/Handlers/Accounts.hs
+++ b/wallet-new/src/Cardano/Wallet/API/V1/Handlers/Accounts.hs
@@ -14,6 +14,8 @@ import           Cardano.Wallet.API.V1.Types
 import           Cardano.Wallet.WalletLayer (PassiveWalletLayer)
 import qualified Cardano.Wallet.WalletLayer as WalletLayer
 
+import           Cardano.Wallet.Kernel.WalletException (throwWalletM)
+
 handlers :: PassiveWalletLayer IO -> ServerT Accounts.API Handler
 handlers w =  deleteAccount w
          :<|> getAccount    w
@@ -30,7 +32,7 @@ deleteAccount :: PassiveWalletLayer IO
 deleteAccount layer wId accIdx = do
     res <- liftIO $ WalletLayer.deleteAccount layer wId accIdx
     case res of
-         Left e   -> throwM e
+         Left e   -> throwWalletM e
          Right () -> return NoContent
 
 -- | Fetches an 'Account' given its parent 'WalletId' and its index.
diff --git a/wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs b/wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs
index 2d82fa6..59cbd98 100644
--- a/wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs
+++ b/wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs
@@ -6,23 +6,17 @@ module Cardano.Wallet.Kernel.Accounts (
     , CreateAccountError(..)
     ) where
 
-import qualified Prelude
 import           Universum
 
-import           Formatting (bprint, build, formatToString, (%))
+import           Formatting (bprint, (%))
 import qualified Formatting as F
 import qualified Formatting.Buildable
 import           System.Random.MWC (GenIO, createSystemRandom, uniformR)
 
-import           Servant (err400)
-
 import           Data.Acid (update)
-import           Data.Aeson (ToJSON, toJSON)
 
 import           Pos.Crypto (EncryptedSecretKey, PassPhrase)
 
-import           Cardano.Wallet.API.V1.Errors (ToHttpErrorStatus (..),
-                     ToServantError (..))
 import           Cardano.Wallet.Kernel.DB.AcidState (CreateHdAccount (..), DB,
                      DeleteHdAccount (..), UpdateHdAccountName (..))
 import           Cardano.Wallet.Kernel.DB.HdWallet (AccountName (..),
@@ -37,8 +31,6 @@ import           Cardano.Wallet.Kernel.Internal (PassiveWallet, walletKeystore,
                      wallets)
 import qualified Cardano.Wallet.Kernel.Keystore as Keystore
 import           Cardano.Wallet.Kernel.Types (WalletId (..))
-import           Cardano.Wallet.Kernel.WalletException
-                     (walletExceptionFromException, walletExceptionToException)
 
 import           Test.QuickCheck (Arbitrary (..), oneof)
 
@@ -65,18 +57,6 @@ instance Buildable CreateAccountError where
     build (CreateAccountHdRndAccountSpaceSaturated hdAcc) =
         bprint ("CreateAccountHdRndAccountSpaceSaturated " % F.build) hdAcc
 
-instance ToJSON CreateAccountError where
-    toJSON = error "TODO"
-instance ToServantError CreateAccountError where
-    declareServantError _ = err400 -- FIXME: picked by random for now !!!!!!!!!!
-instance ToHttpErrorStatus CreateAccountError
-
-instance Show CreateAccountError where
-    show = formatToString build
-
-instance Exception CreateAccountError where
-    toException   = walletExceptionToException
-    fromException = walletExceptionFromException
 
 -- | Creates a new 'Account' for the input wallet.
 -- Note: @it does not@ generate a new 'Address' to go in tandem with this
diff --git a/wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs b/wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs
index 620d54c..6f4dffd 100644
--- a/wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs
+++ b/wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs
@@ -1,4 +1,5 @@
-{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RankNTypes    #-}
 -- TODO: Not sure about the best way to avoid the orphan instances here
 {-# OPTIONS_GHC -fno-warn-orphans #-}
 
@@ -96,6 +97,9 @@ import           Universum
 
 import           Control.Lens (at, (+~), _Wrapped)
 import           Control.Lens.TH (makeLenses)
+import           Data.Aeson (FromJSON (..), ToJSON (..), genericParseJSON,
+                     genericToJSON)
+import qualified Data.Aeson as Aeson
 import qualified Data.ByteString as BS
 import qualified Data.IxSet.Typed as IxSet (Indexable (..))
 import           Data.SafeCopy (SafeCopy (..), base, deriveSafeCopy)
@@ -109,7 +113,7 @@ import qualified Pos.Core as Core
 import           Pos.Core.Chrono (NewestFirst (..))
 import qualified Pos.Crypto as Core
 
-import           Cardano.Wallet.API.V1.Types (V1 (..))
+import           Cardano.Wallet.API.V1.Types (V1 (..), unV1)
 import           Cardano.Wallet.Kernel.DB.InDb
 import           Cardano.Wallet.Kernel.DB.Spec
 import           Cardano.Wallet.Kernel.DB.Util.AcidState
@@ -129,7 +133,13 @@ newtype AccountName = AccountName { getAccountName :: Text }
 
 -- | Account index
 newtype HdAccountIx = HdAccountIx { getHdAccountIx :: Word32 }
-  deriving (Eq, Ord)
+  deriving (Generic, Eq, Ord)
+
+instance ToJSON HdAccountIx where
+    toJSON = genericToJSON Aeson.defaultOptions
+
+instance FromJSON HdAccountIx where
+    parseJSON = genericParseJSON Aeson.defaultOptions
 
 -- NOTE(adn) if we need to generate only @hardened@ account indexes, we
 -- need to extend this arbitrary instance accordingly.
@@ -209,7 +219,7 @@ eskToHdRootId = HdRootId . InDb . Core.makePubKeyAddressBoot . Core.encToPublic
 -- as a primary key. This however is a slightly larger refactoring we don't
 -- currently have time for.
 newtype HdRootId = HdRootId { getHdRootId :: InDb Core.Address }
-  deriving (Eq, Ord)
+  deriving (Generic, Eq, Ord)
 
 instance Arbitrary HdRootId where
   arbitrary = do
@@ -217,12 +227,18 @@ instance Arbitrary HdRootId where
                                                <*> pure mempty
       pure (eskToHdRootId esk)
 
+instance ToJSON HdRootId where
+    toJSON = genericToJSON Aeson.defaultOptions
+
+instance FromJSON HdRootId where
+    parseJSON = genericParseJSON Aeson.defaultOptions
+
 -- | HD wallet account ID
 data HdAccountId = HdAccountId {
       _hdAccountIdParent :: HdRootId
     , _hdAccountIdIx     :: HdAccountIx
     }
-  deriving (Eq)
+  deriving (Generic, Eq)
 
 -- | We make sure to compare the account index first to avoid doing an
 -- unnecessary comparison of the root ID
@@ -234,6 +250,12 @@ instance Ord HdAccountId where
 instance Arbitrary HdAccountId where
   arbitrary = HdAccountId <$> arbitrary <*> arbitrary
 
+instance ToJSON HdAccountId where
+    toJSON = genericToJSON Aeson.defaultOptions
+
+instance FromJSON HdAccountId where
+    parseJSON = genericParseJSON Aeson.defaultOptions
+
 -- | HD wallet address ID
 data HdAddressId = HdAddressId {
       _hdAddressIdParent :: HdAccountId
@@ -473,13 +495,21 @@ data UnknownHdAccount =
 
     -- | Unknown account (implies the root is known)
   | UnknownHdAccount HdAccountId
-  deriving Eq
+  deriving (Generic, Eq)
 
 instance Arbitrary UnknownHdAccount where
     arbitrary = oneof [ UnknownHdAccountRoot <$> arbitrary
                       , UnknownHdAccount <$> arbitrary
                       ]
 
+instance ToJSON (V1 UnknownHdAccount) where
+    toJSON = genericToJSON Aeson.defaultOptions . unV1
+
+instance FromJSON (V1 UnknownHdAccount) where
+    parseJSON = fmap V1 . genericParseJSON Aeson.defaultOptions
+
+
+
 -- | Unknown address
 data UnknownHdAddress =
     -- | Unknown root ID
diff --git a/wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs b/wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs
index df02c38..54374b9 100644
--- a/wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs
+++ b/wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs
@@ -8,6 +8,7 @@ module Cardano.Wallet.Kernel.DB.InDb (
 import           Universum
 
 import           Control.Lens.TH (makeLenses)
+import           Data.Aeson (FromJSON (..), ToJSON (..))
 import           Data.SafeCopy (SafeCopy (..))
 
 import qualified Pos.Chain.Block as Core
@@ -30,6 +31,12 @@ instance Applicative InDb where
   pure = InDb
   InDb f <*> InDb x = InDb (f x)
 
+instance ToJSON a => ToJSON (InDb a) where
+    toJSON = toJSON . _fromDb
+
+instance FromJSON a => FromJSON (InDb a) where
+    parseJSON = fmap InDb . parseJSON
+
 makeLenses ''InDb
 
 {-------------------------------------------------------------------------------
diff --git a/wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs b/wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs
index 2c2f3c2..51466cf 100644
--- a/wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs
+++ b/wallet-new/src/Cardano/Wallet/Kernel/WalletException.hs
@@ -1,42 +1,35 @@
 {-# LANGUAGE ExistentialQuantification #-}
 
-module Cardano.Wallet.Kernel.WalletException where
+module Cardano.Wallet.Kernel.WalletException
+    ( WalletException(..)
+    , throwWalletM
+    ) where
 
 import           Universum
 
-import           Data.Aeson (ToJSON, toJSON)
-import           Data.Typeable (cast)
-import           Formatting.Buildable (build)
-import           Prelude (show)
+import           Cardano.Wallet.API.Response.JSend (HasDiagnostic)
+import           Cardano.Wallet.API.V1.Errors (ToHttpErrorStatus)
 
-import           Cardano.Wallet.API.V1.Errors (ToHttpErrorStatus (..),
-                     ToServantError (..))
+import qualified Prelude
 
-data WalletException = forall e . (ToHttpErrorStatus e, Exception e, Buildable e) => WalletException e
 
--- FIXME: Generics somewhow?
+data WalletException = forall e.
+    ( Exception e
+    , HasDiagnostic e
+    , ToHttpErrorStatus e
+    ) => WalletException { getWalletException :: e }
 
-instance Show WalletException where
+instance Prelude.Show WalletException where
     show (WalletException e) = Prelude.show e
 
-instance ToJSON WalletException where
-    toJSON (WalletException e) = toJSON e
-
-instance ToHttpErrorStatus WalletException where
-    toHttpErrorStatus (WalletException e) = toHttpErrorStatus e
-
-instance ToServantError WalletException where
-    declareServantError (WalletException e) = declareServantError e
-
-instance Buildable WalletException where
-    build (WalletException e) = build e
-
 instance Exception WalletException
 
-walletExceptionToException :: (ToHttpErrorStatus e, Exception e) => e -> SomeException
-walletExceptionToException = toException . WalletException
-
-walletExceptionFromException :: Exception e => SomeException -> Maybe e
-walletExceptionFromException x = do
-    WalletException a <- fromException x
-    cast a
+-- | Wrap a wallet exception in a proper type that can be caught later
+throwWalletM ::
+    ( Exception e
+    , MonadThrow m
+    , HasDiagnostic e
+    , ToHttpErrorStatus e
+    ) => e -> m a
+throwWalletM =
+    throwM . WalletException
diff --git a/wallet-new/src/Cardano/Wallet/Server/Plugins.hs b/wallet-new/src/Cardano/Wallet/Server/Plugins.hs
index d391b56..51f6aec 100644
--- a/wallet-new/src/Cardano/Wallet/Server/Plugins.hs
+++ b/wallet-new/src/Cardano/Wallet/Server/Plugins.hs
@@ -24,7 +24,7 @@ import qualified Cardano.Wallet.API.V1.Types as V1
 import           Cardano.Wallet.Kernel (PassiveWallet)
 import qualified Cardano.Wallet.Kernel.Diffusion as Kernel
 import qualified Cardano.Wallet.Kernel.Mode as Kernel
-import           Cardano.Wallet.Kernel.WalletException (WalletException)
+import           Cardano.Wallet.Kernel.WalletException (WalletException (..))
 import qualified Cardano.Wallet.LegacyServer as LegacyServer
 import qualified Cardano.Wallet.Server as Server
 import           Cardano.Wallet.Server.CLI (NewWalletBackendParams (..),
@@ -162,7 +162,7 @@ legacyWalletBackend pm txpConfig WalletBackendParams {..} ntpStatus = pure $ \di
 
     exceptionHandler :: SomeException -> Response
     exceptionHandler se =
-        case asum [handleV1Errors se, handleWalletException <$> fromException se, handleV0Errors se] of
+        case asum [handleV1Errors se, handleV0Errors se] of
              Nothing -> handleGenericError se
              Just r  -> r
 
@@ -173,10 +173,6 @@ legacyWalletBackend pm txpConfig WalletBackendParams {..} ntpStatus = pure $ \di
                 responseLBS (V1.toHttpErrorStatus we) [applicationJson] .  encode $ we
         in fmap reify (fromException se)
 
-    handleWalletException :: WalletException -> Response
-    handleWalletException e =
-          responseLBS (V1.toHttpErrorStatus e) [applicationJson] .  encode $ e
-
     -- Handles domain-specific errors coming from the V0 API, but rewraps it
     -- into a jsend payload. It doesn't explicitly handle 'InternalError' or
     -- 'DecodeError', as they can come from any part of the stack or even
@@ -236,6 +232,11 @@ walletBackend protocolMagic (NewWalletBackendParams WalletBackendParams{..}) (pa
     lower :: env -> ReaderT env IO a -> IO a
     lower env m = runReaderT m env
 
+    handleWalletException :: WalletException -> Response
+    handleWalletException (WalletException e) =
+          responseLBS (V1.toHttpErrorStatus e) [applicationJson] . encode $ e
+
+
 -- | A @Plugin@ to resubmit pending transactions.
 resubmitterPlugin :: HasConfigurations
                   => ProtocolMagic
diff --git a/wallet-new/src/Cardano/Wallet/WalletLayer.hs b/wallet-new/src/Cardano/Wallet/WalletLayer.hs
index 09a7f16..bbb8558 100644
--- a/wallet-new/src/Cardano/Wallet/WalletLayer.hs
+++ b/wallet-new/src/Cardano/Wallet/WalletLayer.hs
@@ -1,3 +1,6 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase    #-}
+
 module Cardano.Wallet.WalletLayer
     ( PassiveWalletLayer (..)
     , ActiveWalletLayer (..)
@@ -22,9 +25,12 @@ module Cardano.Wallet.WalletLayer
 
 import           Universum
 
+import           Data.Aeson (FromJSON (..), ToJSON (..))
 import           Formatting (bprint, build, formatToString, (%))
 import qualified Formatting.Buildable
+import           Generics.SOP.TH (deriveGeneric)
 import qualified Prelude
+import           Servant (err400, err404)
 import           Test.QuickCheck (Arbitrary (..), oneof)
 
 import           Pos.Chain.Block (Blund)
@@ -37,6 +43,11 @@ import           Cardano.Wallet.API.Request (RequestParams (..))
 import           Cardano.Wallet.API.Request.Filter (FilterOperations (..))
 import           Cardano.Wallet.API.Request.Sort (SortOperations (..))
 import           Cardano.Wallet.API.Response (SliceOf (..), WalletResponse)
+import           Cardano.Wallet.API.Response.JSend (HasDiagnostic (..))
+import           Cardano.Wallet.API.V1.Errors (ToHttpErrorStatus,
+                     ToServantError (..))
+import           Cardano.Wallet.API.V1.Generic (jsendErrorGenericParseJSON,
+                     jsendErrorGenericToJSON)
 import           Cardano.Wallet.API.V1.Types (Account, AccountBalance,
                      AccountIndex, AccountUpdate, Address, NewAccount,
                      NewAddress, NewWallet, NodeInfo, NodeSettings,
@@ -247,20 +258,40 @@ instance Buildable GetAccountError where
 data DeleteAccountError =
       DeleteAccountError (V1 Kernel.UnknownHdAccount)
     | DeleteAccountWalletIdDecodingFailed Text
-    deriving Eq
+    deriving (Generic, Eq)
+
+deriveGeneric ''DeleteAccountError
 
 -- | Unsound show instance needed for the 'Exception' instance.
 instance Show DeleteAccountError where
     show = formatToString build
 
-instance Exception DeleteAccountError
-
 instance Buildable DeleteAccountError where
     build (DeleteAccountError kernelError) =
         bprint ("DeleteAccountError " % build) kernelError
     build (DeleteAccountWalletIdDecodingFailed txt) =
         bprint ("DeleteAccountWalletIdDecodingFailed " % build) txt
 
+instance HasDiagnostic DeleteAccountError where
+    getDiagnosticKey (DeleteAccountError _) = "unknownAccount"
+    getDiagnosticKey (DeleteAccountWalletIdDecodingFailed _) = "message"
+
+instance ToServantError DeleteAccountError where
+    declareServantError = \case
+        DeleteAccountError _ -> err404
+        DeleteAccountWalletIdDecodingFailed _ -> err400
+
+instance ToHttpErrorStatus DeleteAccountError
+
+instance ToJSON DeleteAccountError where
+    toJSON = jsendErrorGenericToJSON
+
+instance FromJSON DeleteAccountError where
+    parseJSON = jsendErrorGenericParseJSON
+
+instance Exception DeleteAccountError
+
+
 data GetAccountsError =
       GetAccountsError Kernel.UnknownHdRoot
     | GetAccountsWalletIdDecodingFailed Text
```

KtorZ commented:
What we are basically missing is all the `HasDiagnostic`, `ToServantError`, `ToJSON` and `FromJSON` instances for all errors and their inherent types. 

Also, we will need to define a set of roundtrips Quickcheck tests for each errors the same we we do it here:

- https://github.com/input-output-hk/cardano-sl/blob/ce0e2321a6d38363a590ccf3f7451e813315bdb2/wallet-new/test/MarshallingSpec.hs#L57-L59

and ideally, a set of golden tests as defined here:

- https://github.com/input-output-hk/cardano-sl/blob/ce0e2321a6d38363a590ccf3f7451e813315bdb2/wallet-new/test/WalletNewJson.hs

parsonsmatt commented on wallet-new/src/Cardano/Wallet/Kernel/DB/InDb.hs:
```diff
@@ -30,6 +31,12 @@ instance Applicative InDb where
+    toJSON = toJSON . _fromDb
+
+instance FromJSON a => FromJSON (InDb a) where
+    parseJSON = fmap InDb . parseJSON
```
These instances are the same you'd get from `GeneralizedNewtypeDeriving`

parsonsmatt commented on wallet-new/src/Cardano/Wallet/WalletLayer/Exception.hs:
```diff
@@ -0,0 +1,35 @@
+    , ToHttpErrorStatus e
+    ) => e -> m a
+throwWalletM =
+    throwM . WalletException
```
We shouldn't need this. Instead we should use `throwM` directly, and have the `Exception` instances for our types wrap in `WalletException`.

Anviking commented on wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs:
```diff
@@ -49,10 +52,17 @@ data CreateAddressError =
 
 instance Arbitrary CreateAddressError where
-    arbitrary = oneof []
+    arbitrary = oneof [ CreateAddressUnknownHdAccount <$> arbitrary
```
All of these `Arbitrary` instances could be replaced with `genericArbitrary`. Would rather move fast with this pr first though.

I think `genericArbitrary` would be really nice though, as a small amount of existing instances looked incomplete, as in "let's get this compiling", and not "this is exactly the arbitrary instance we need".

Anviking commented on wallet-new/src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -41,6 +46,7 @@ import qualified Cardano.Wallet.Kernel.Diffusion as Kernel
 import qualified Cardano.Wallet.Kernel.Mode as Kernel
 import qualified Cardano.Wallet.Server as Server
 import qualified Cardano.Wallet.WalletLayer.Kernel as WalletLayer.Kernel
+import           Data.Aeson (encode)
```
Didn't get how these imports are grouped.

KtorZ replied:
Unfortunately, there's no way to enforce rules on imports, that would require the linter to have some semantics notions about them. So, the best we can do is to enforce this -- which also gets complicated in merges and rebases :| 

Please, follow as much as possible @erikd's suggestion on that.

jcmincke commented:
@KtorZ @parsonsmatt : could you please (continue to) review this PR?

erikd commented:
This needs rebasing and its probably also a good idea to squash all those fixup commits.

KtorZ commented on wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs:
```diff
@@ -211,20 +222,23 @@ eskToHdRootId = HdRootId . InDb . Core.makePubKeyAddressBoot . Core.encToPublic
 instance Arbitrary HdRootId where
   arbitrary = do
-      (_, esk) <- Core.safeDeterministicKeyGen <$> (BS.pack <$> vectorOf 12 arbitrary)
+      (_, esk) <- Core.safeDeterministicKeyGen <$> (BS.pack <$> vectorOf 32 arbitrary)
```
Why exactly :) ?

Anviking replied:
Aeson roundtrips didn't work because of this. 
```
Exception thrown while showing test case:
           Wallet.generate: seed needs to be >= 32 bytes, got : 12
```
(06c8d9424010c5909dd8f8bb8544623b83cd4cd2)

KtorZ replied:
:+1:

KtorZ commented on wallet-new/src/Cardano/Wallet/WalletLayer.hs:
```diff
@@ -1,3 +1,6 @@
+{-# LANGUAGE DeriveGeneric #-}
```
Better to leave that file untouched for now. Ryan can most probably hook up all the errors in the old data structure without all the instances here. 
I apologize for the work done here, I've a bit overlooked stuff.

KtorZ commented:
Side note:

This PR shouldn't be longer than one or two commits. So please, eventually rebase and squash stuff in a meaningful way :) 
If you need some help / advice with that, let me know!

Anviking commented on wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -208,13 +212,20 @@ data InvalidRedemptionCode =
 
     -- | AES decryption error (for paper wallets)
-  | InvalidRedemptionCodeCryptoError CryptoError
+  | InvalidRedemptionCodeCryptoError Text
```
To avoid orphan instances for the `cryptonite` `CryptoError`… which is perhaps unnecessarily ideological as we're giving up functionality?
https://input-output-rnd.slack.com/archives/G8PHF04B0/p1535541556000100

Similar thing going on with [TimeExecutionLimit](https://github.com/input-output-hk/cardano-sl/blob/de6e30deb05a70c7aed682a9123c9e8bd9a14422/wallet-new/src/Cardano/Wallet/WalletLayer/ExecutionTimeLimit.hs#L24-L35), but there I did custom instances.

Anviking replied:
Borderline similar thing going in with #3493 too, but that one uses internal types and is more tricky.

KtorZ replied:
Hmmmm.. Okay.

Anviking commented on wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -233,6 +244,11 @@ instance Buildable InvalidRedemptionCode where
 
+instance Arbitrary InvalidRedemptionCode where
+    arbitrary = oneof [ InvalidRedemptionCodeInvalidBase64 <$> arbitrary
+                      , InvalidRedemptionCodeCryptoError <$> arbitrary
```
Looking into this. Think we should have an arbitrary instance for `CryptoError`, even if they are turned into strings (which I'm also beginning to question more).

This is just using the `arbitrary` of `String`.

Anviking replied:
Done. Pushing soon. Also, realised `CryptoError` has no payload, so `Text` seems fine to me.

KtorZ replied:
Sorry, I was busy and didn't answer. To answer you: I'd rather see a newtype than an orphan instance :) 
You could use a newtype locally to wrap and unwrap the error in order to define the Arbitrary instance.

Anviking replied:
Oh no, sorry, was mostly talking out loudly, and also formulated myself carelessly.

I simply wrote a `genCryptoError :: Gen CryptoError` helper. 

(2456a74ef2fa4c59584187938552f5219cca9976, rebasing soon)

Anviking commented:
@KtorZ should be addressed now

Anviking commented on wallet-new/src/Cardano/Wallet/WalletLayer.hs:
```diff
@@ -367,15 +352,71 @@ instance Buildable GetTxError where
+    build (RedeemAdaWalletIdDecodingFailed txt) =
+        bprint ("RedeemAdaWalletIdDecodingFailed " % build) txt
+    build (RedeemAdaInvalidRedemptionCode txt) =
+        bprint ("RedeemAdaInvalidRedemptionCode " % build) txt
```
5352961fc2e86a9536f44bcd0cfaed5f706565e6 this commit? It just swaps two sections. We could drop it, but I found it needed for implementing instances for the `WalletLayer` errors to avoid Template Haskell problems.

KtorZ replied:
Okay :/ .. But the template haskell was only for the SOP-Generics right? Since we dropped it, I wouldn't do the swap either here and keep the PR just about the rest.

Anviking commented on wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs:
```diff
@@ -133,7 +137,10 @@ newtype AccountName = AccountName { getAccountName :: Text }
+  deriving (Generic, Eq, Ord, Show)
+
+instance ToJSON HdAccountIx
+instance FromJSON HdAccountIx
```
Ah, thanks for explaining.

Anviking commented on wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Conv.hs:
```diff
@@ -213,13 +218,20 @@ data InvalidRedemptionCode =
 
     -- | AES decryption error (for paper wallets)
-  | InvalidRedemptionCodeCryptoError CryptoError
+  | InvalidRedemptionCodeCryptoError Text
```
We would need custom `toJSON`/`parseJSON` functions then.

KtorZ replied:
Well, in a similar fashion to what's done for the arbitrary generator I suppose? There''s probably a `Read` and `Show` instance on `CryptoError`, so we can build up a parser using `Safe.readMaybe` ?

Anviking replied:
Only `(Show,Eq,Enum,Data,Typeable)` and `Exception`. Can we do something else with that?

KtorZ replied:
Ah. Erf. Yeah, but that becomes a bit dodgy. Let's go for `Text` here :/

adinapoli-iohk commented on wallet-new/src/Cardano/Wallet/Kernel/DB/HdWallet.hs:
```diff
@@ -485,13 +508,19 @@ data UnknownHdAccount =
+deriving newtype instance FromJSON (V1 UnknownHdAccount)
+instance ToJSON UnknownHdAccount
+instance FromJSON UnknownHdAccount
+
```
Ah, yes, good question, @ouroboros spotted the inconsistency as well.

There is a rationale behind it, though. We decided to do this with @edsko a while ago to signal that this is an error reporting behaviour specific to `V1`, which might change in the future (for example if we plug external wallets or non-hd wallets in the future). It would allow us to do the same trick of deriving particular instances for `V1 Foo` rather than just `Foo`, which might be handy.

But you are right, we should do this consistently across the board. That's an oversight, oops 😉

Anviking commented:
I removed the commit adding `WalletLayer.Exception` now since 1. I feel it's separate and 2. I'm not at all convinced it's the correct way.

Also, I have removed the instances for `(V1 SomeType)` since they shouldn't be needed any more.
